includes/class-meta-pixel.php

```diff
@@ -0,0 +1,228 @@
+	 * @return boolean
+	 */
+	public static function validate_pixel_id( $value ) {
+		return '' === $value || ctype_digit( $value ) || is_int( $value );
+	}
+
+	/**
+	 * Callback for the REST API GET method
+	 *
+	 * @return \WP_REST_Response
+	 */
+	public static function api_get() {
+		return rest_ensure_response( self::get_option() );
+	}
+
+	/**
+	 * Callback for the REST API POST method to save the settings
+	 *
+	 * @param \WP_REST_Request $request The request object.
+	 * @return \WP_REST_Response
+	 */
+	public static function api_save( $request ) {
+		$pixel_id = $request->get_param( 'pixel_id' );
+		$active   = $request->get_param( 'active' );
+		$value    = compact( 'pixel_id', 'active' );
+		update_option( self::OPTION_NAME, self::sanitize_option( $value ) );
+		return rest_ensure_response( self::get_option(), 200 );
+	}
+
+	/**
+	 * Gets the current value of the option
+	 *
+	 * @return array
+	 */
+	public static function get_option() {
+		return get_option( self::OPTION_NAME, self::get_default_values() );
+	}
+
+	/**
+	 * Checks if Meta pixel option is active
+	 *
+	 * @return boolean
+	 */
+	public static function is_active() {
+		return self::get_option()['active'];
+	}
+
+	/**
+	 * Gets the stored pixel ID. If Meta Pixel option is not active, ignore the saved option
+	 *
+	 * @return string
+	 */
+	public static function get_pixel_id() {
+		if ( ! self::is_active() ) {
+			return '';
+		}
+		return self::get_option()['pixel_id'];
+	}
+
+	/**
+	 * Get settings default values
+	 *
+	 * @return array
+	 */
+	public static function get_default_values() {
+		return [
+			'active'   => false,
+			'pixel_id' => '',
+		];
+	}
+
+	/**
+	 * Sanitizes settings
+	 *
+	 * @param array $value The settings value.
+	 * @return array
+	 */
+	public static function sanitize_option( $value ) {
+		$defaults  = self::get_default_values();
+		$sanitized = [];
+		if ( ! is_array( $value ) ) {
+			return $defaults;
+		}
+		if ( isset( $value['active'] ) ) {
+			$sanitized['active'] = (bool) $value['active'];
+		} else {
+			$sanitized['active'] = $defaults['active'];
+		}
+		if ( isset( $value['pixel_id'] ) ) {
+			$sanitized['pixel_id'] = (string) $value['pixel_id']; // store as string to avoid PHP_INT_MAX problems.
+		} else {
+			$sanitized['pixel_id'] = $defaults['pixel_id'];
+		}
+		return $sanitized;
+	}
+
+	/**
+	 * Checks if AMP plugin is enabled and if the current response is being server as AMP
+	 *
+	 * @return boolean
+	 */
+	public static function is_amp() {
+		return function_exists( 'is_amp_endpoint' ) && is_amp_endpoint();
+	}
+
+	/**
+	 * Gets the template for the img tag snippet
+	 *
+	 * @return string
+	 */
+	public static function get_img_snippet() {
+		global $wp;
+		$current_user = wp_get_current_user();
+		$event_params = [
+			'page_title' => get_the_title(),
+			'user_role'  => empty( $current_user->roles ) ? 'guest' : $current_user->roles[0],
+			'event_url'  => home_url( $wp->request ),
+		];
```

Comment:
Are these values standard? Looks like the parameters being used for the img snippet are not being applied to the script. Won't this cause inconsistent tracking?

It would be nice to have them filterable in an isolated private method like `get_snippet_parameters()` to be used in both img and script snippets.

Reply:
All examples I've seen of the JS snippet do not include these vars...

Reply:
Added the custom properties to the JS snippet following what I see here:

https://developers.facebook.com/docs/meta-pixel/implementation/conversion-tracking#custom-properties

Or we could also remove them from both places... The request in the [original issue](https://github.com/Automattic/newspack-plugin/issues/1554) did not ask for these properties.

I think this is something I brought looking at one of the plugins that add the pixel.

Reply:
I believe it's good practice to implement a default set of parameters and make it filterable! This is working well 👍